packages/fxa-auth-server/lib/routes/attached-clients.js

```diff
@@ -0,0 +1,264 @@
+          client.os = session.uaOS;
+          if (! session.uaBrowser) {
+            client.userAgent = '';
+          } else if (! session.uaBrowserVersion) {
+            client.userAgent = session.uaBrowser;
+          } else {
+            const { uaBrowser: browser, uaBrowserVersion: version } = session;
+            client.userAgent = `${browser} ${version.split('.')[0]}`;
+          }
+        }
+
+        // Merge with OAuth clients, which may or may not be linked
+        // to a device record.
+        for (const oauthClient in await oauthClientsP) {
+          let client = clientsBySessionTokenId.get(oauthClient.refresh_token_id);
+          if (! client) {
+            client = {
+              ...defaultFields,
+              refreshTokenId: oauthClient.refresh_token_id,
+              createdTime: oauthClient.creation_time,
+            };
+            attachedClients.push(client);
+          }
+          client.clientId = oauthClient.client_id;
+          client.createdTime = Math.min(client.createdTime, oauthClient.createdTime);
+          client.lastAccessTime = Math.max(client.lastAccessTime, oauthClient.lastAccessTime);
+        }
+
+        // Now we can do some final tweaks of each item for display.
+        for (const client of attachedClients) {
+          formatTimestamps(client, request);
+          formatLocation(client, request);
+        }
+
+        return attachedClients;
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/attached_client/destroy',
+      options: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: isA.object({
+            clientId: validators.clientId.allow(null).optional(),
+            sessionTokenId: isA.string().regex(HEX_STRING).allow(null).optional(),
+            refreshTokenId: validators.refreshToken.allow(null).optional(),
+            deviceId: DEVICES_SCHEMA.id.allow(null).optional(),
+          })
+          .or('clientId', 'sessionTokenId', 'refreshTokenId', 'deviceId')
+          .with('refreshTokenId', ['clientId'])
+        },
+        response: {
+          schema: {}
+        }
+      },
+      handler: async function (request) {
+        log.begin('Account.attachedClientDestroy', request);
+
+        const sessionToken = request.auth.credentials;
+        const uid = sessionToken.uid;
+
+        if (request.payload.deviceId) {
+          // If we got a `deviceId`, then deleting that should also delete the sessionToken
+          // and should return the refreshToken. It's fishy if the values returned here don't
+          // match ones we were given in the request.
+          // XXX TODO: needs to trigger notifications, like `/account/device/destroy`.
+          const device = await db.deleteDevice(uid, request.payload.deviceId);
+          if (device && device.refreshTokenId) {
+            await oauthdb.revokeRefreshTokenById(device.refreshTokenId);
```

Comment:
From mtg: this will also need to wipe access tokens, to prevent the client from showing up in the "devices and apps" list.